### PR TITLE
[AIRFLOW-3787] Import/export users from JSON file

### PR DESCRIPTION
**Pull request by astahlman**
Make sure you have checked _all_ steps below.

### Jira

- [ X ] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3787
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [ X ] Here are some details about my PR, including screenshots of any UI changes:

Provide a CLI command to import or export users from a JSON file. The
CLI arguments are modeled after the import/export commands for Variables
and Pools.

Example Usage:

    airflow users -i users.json
    airflow users -e /tmp/exported-users.json

The import command will create any users that do not yet exist and
update any users that already exist. It never deletes users.

The format of the file produced by an export is compatible with the
import command, i.e., `import(export())` should succeed but have no
side-effects.

### Tests

- [ X ] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

Added unit tests for both import and export.

### Commits

- [ X ] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.
  - All the public functions and the classes in the PR contain docstrings that explain what it does

### Code Quality

- [ X ] Passes `flake8`

### CODE REVIEW

**Comment by astahlman** [comment]
cc @feng-tao

File: airflow/bin/cli.py [hunk 1]
```diff
@@ -1474,6 +1474,107 @@ def users(args):
                 print('User "{}" added to role "{}".'.format(
                     user,
                     args.role))
+    elif args.export:
+        appbuilder = cached_appbuilder()
+        users = appbuilder.sm.get_all_users()
+        fields = ['id', 'username', 'email', 'first_name', 'last_name', 'roles']
+
+        # In the User model the first and last name fields have underscores,
+        # but the corresponding parameters in the CLI don't
```
Comment (feng-tao) [inline airflow/bin/cli.py:1484]: I am not sure if I understand the comment, could you clarify which part of cli doesn't allow the underscore?

File: airflow/bin/cli.py [hunk 2]
```diff
@@ -1474,6 +1474,107 @@ def users(args):
...
+             if field != 'roles' else [r.name for r in user.roles]
+             for field in fields}
+            for user in users
+        ]
+
+        with open(args.export, 'w') as f:
+            f.write(json.dumps(users, sort_keys=True, indent=4))
+            print("{} users successfully exported to {}".format(len(users), f.name))
+
+    elif getattr(args, 'import'):  # "import" is a reserved word
```
Comment (feng-tao) [inline airflow/bin/cli.py:1499]: so ``args.import`` will throw an exception in this case?

File: airflow/bin/cli.py [hunk 3]
```diff
@@ -1474,6 +1474,107 @@ def users(args):
...
+        roles = []
+        for rolename in user['roles']:
+            role = appbuilder.sm.find_role(rolename)
+            if not role:
+                print("Error: '{}' is not a valid role".format(rolename))
+                exit(1)
+            else:
+                roles.append(role)
+
+        required_fields = ['username', 'firstname', 'lastname',
```
Comment (feng-tao) [inline airflow/bin/cli.py:1538]: ah, I see now. is there any reason we don't use first_name, last_name for the field name when import from a file?

File: airflow/bin/cli.py [hunk 4]
```diff
@@ -1474,6 +1474,107 @@ def users(args):
...
+            appbuilder.sm.update_user(existing_user)
+            users_updated.append(user['email'])
+        else:
+            print("Creating new user with email '{}'".format(user['email']))
+            appbuilder.sm.add_user(
+                username=user['username'],
+                first_name=user['firstname'],
+                last_name=user['lastname'],
+                email=user['email'],
+                role=roles[0],  # add_user() requires exactly 1 role
```
Comment (feng-tao) [inline airflow/bin/cli.py:1568]: this may be worth to update FAB. but we could do it later.

File: airflow/bin/cli.py [hunk 5]
```diff
@@ -1474,6 +1474,107 @@ def users(args):
...
+            for user in users
+        ]
+
+        with open(args.export, 'w') as f:
+            f.write(json.dumps(users, sort_keys=True, indent=4))
+            print("{} users successfully exported to {}".format(len(users), f.name))
+
+    elif getattr(args, 'import'):  # "import" is a reserved word
+        json_file = getattr(args, 'import')
+        if not os.path.exists(json_file):
```
Comment (feng-tao) [inline airflow/bin/cli.py:1501]: it would be good to give an example what an import json file will look like.

**Review by feng-tao — COMMENTED** [review]
Thanks @astahlman . The code LGTM overall. But couples of questions(may not directly relate to this pr):
1. Could we configure a default role for any new user who has never login before in FAB? Or it has to run ``airflow users -c`` first? Is there any different behavior when we use different authentication type(e.g LDAP vs OATH)?
2. Is there a way to limit the cli to run by Admin only? I think from security point of view, we should think about this if it doesn't support(could be a different pr).

File: airflow/bin/cli.py (hunk 2 above)
Comment (astahlman) [reply airflow/bin/cli.py:1499]: Yep, it's a SyntaxError:

```
ERROR: Failure: SyntaxError (invalid syntax (cli.py, line 1498))
```

File: airflow/bin/cli.py (hunk 3 above)
Comment (astahlman) [reply airflow/bin/cli.py:1538]: We could, but I thought it made sense to be consistent with the CLI, which takes `--firstname` and `--lastname`

File: airflow/bin/cli.py (hunk 5 above)
Comment (astahlman) [reply airflow/bin/cli.py:1501]: Sure, where's the best place for that? Here, as a comment, or are there some docs for the CLI?

File: airflow/bin/cli.py (hunk 3 above)
Comment (feng-tao) [reply airflow/bin/cli.py:1538]: I see. In this case, let's be consistent with the CLI then. Thanks for the clarification.

File: airflow/bin/cli.py (hunk 5 above)
Comment (feng-tao) [reply airflow/bin/cli.py:1501]: do you think it will be too much to put it in the help comment for import option?

File: airflow/bin/cli.py [hunk 6]
```diff
@@ -1998,6 +2099,14 @@ class CLIFactory(object):
             ('--remove-role',),
             help='Remove user from a role',
             action='store_true'),
+        'user_import': Arg(
+            ("-i", "--import"),
+            metavar="FILEPATH",
+            help="Import users from JSON file"),
```
Comment (feng-tao) [inline airflow/bin/cli.py]: what do you think for putting the example file format here?

**Comment by astahlman** [comment]
@feng-tao Answers inline:

> Could we configure a default role for any new user who has never login before in FAB? Or it has to run airflow users -c first?

You don't have to run `users -c` first - the import command will create any users that don't already exist with the `roles` configured in the .json file. I'm not sure how we would use a default, since the `role` field has to be supplied in the .json configuration.

> Is there any different behavior when we use different authentication type(e.g LDAP vs OATH)?

The behavior of the command should be the same, but the way it's used would be different. I think the ideal process for OAuth is:

1.  User registers via the UI, which creates the user in the DB with their username set to the ID used by the identity provider (for Google OAuth, this is a long ID string)
2. An admin updates `<file>.json` to add the user to the appropriate roles
3. Admin (or some automated process) runs `airflow -i <file>.json` to ensure all users are assigned to the correct roles

> Is there a way to limit the cli to run by Admin only? I think from security point of view, we should think about this if it doesn't support(could be a different pr).

To run this command successfully, the user would need access to the CLI and read-write permissions on the metastore. I think anyone who has both of these should be considered an administrator - right?

File: airflow/bin/cli.py (hunk 6 above)
Comment (astahlman) [reply airflow/bin/cli.py]: Sure, I can do that - will update the PR shortly

**Comment by feng-tao** [comment]
@astahlman , make sense. One question I may not be clear is actually related to new user registration. For example, whether we could configure a role for a new user for his first login if that user hasn't been created before. It seems we could do that through some configure in FAB(https://flask-appbuilder.readthedocs.io/en/latest/user_registration.html). 

To brainstorm the ideal Admin workflow, we should configure a default role for the new user. Then if we want to do bulk update on the user roles, we could use -import option in this case.

**Comment by feng-tao** [comment]
@astahlman , feel free to ping me again once you update the pr.

**Comment by astahlman** [comment]
@feng-tao Updated the PR with the help text.

re: the default role for newly registered users, that's configurable via the `AUTH_USER_REGISTRATION_ROLE` setting: https://github.com/apache/airflow/blob/7ebecd677d427e90e292cb7446d96227bd0d6ae1/airflow/config_templates/default_webserver_config.py#L56-L60

**Comment by feng-tao** [comment]
yeap @astahlman

**Comment by feng-tao** [comment]
will merge once CI passes.

**Pull request merged**